### PR TITLE
[Catalog] Implement API for setting app-wide theming

### PR DESCRIPTION
**Pull request by yarneo**
Pivotal Story: https://www.pivotaltracker.com/story/show/156616695

This includes an implementation example in the ButtonTypicalUse example. In a separate PR will this be added to the rest of our examples.

### CODE REVIEW

File: catalog/MDCCatalog/MDCNodeListViewController.swift [hunk 1]
```diff
@@ -165,13 +165,31 @@ class MDCNodeListViewController: CBCNodeListViewController {
...
+  func applyThemeOnCurrentExample() {
+    // This is a short term solution of applying the theme instantly on the
+    // currently presented example. A better solution would be to re-theme the example's components
+    // and not reload the example entirely with a pop-push.
+    guard let topVC = self.navigationController?.topViewController,
+      topVC is MDCAppBarContainerViewController ||
+        topVC.self.responds(to: NSSelectorFromString("catalogBreadcrumbs")) else {
+          return
+    }
+    guard self.navigationController!.popViewController(animated: false) != nil,
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: Try to keep one statement per guard for legibility's-sake.

E.g.:

```swift
guard let navigationController = self.navigationController else {
  return
}
```

File: catalog/MDCCatalog/MDCNodeListViewController.swift [hunk 2]
```diff
@@ -395,54 +413,69 @@ extension MDCNodeListViewController {
...
-      }
+      selectedNode = node
+      vc = setupExample()
     } else {
       vc = MDCNodeListViewController(node: node)
     }
     self.navigationController?.pushViewController(vc, animated: true)
   }
+
+  func setupExample() -> UIViewController {
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: Consider:

```swift
func createViewController(from node: CBCNode) -> UIViewController
```

File: components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.h [hunk 3]
```diff
@@ -21,10 +21,16 @@
 
 #import <UIKit/UIKit.h>
 
+@class MDCSemanticColorScheme;
+@class MDCTypographyScheme;
```
Comment (jverkoey) [inline components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.h]: Import the scheme headers:

```objc
#import "MaterialColorScheme.h"
#import "MaterialTypographyScheme.h"
```

**Review by jverkoey — CHANGES_REQUESTED** [review]
Swift style feedback, otherwise structurally looks good.

File: catalog/MDCCatalog/MDCNodeListViewController.swift [hunk 4]
```diff
@@ -165,13 +165,31 @@ class MDCNodeListViewController: CBCNodeListViewController {
...
+    applyThemeOnCurrentExample()
   }
 
   private func applyColorScheme(_ colorScheme: MDCColorScheming) {
     MDCAppBarColorThemer.applySemanticColorScheme(colorScheme, to: appBar)
 
     appBar.navigationBar.tintColor = UIColor.white
   }
+
+  func applyThemeOnCurrentExample() {
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: `applyThemeToCurrentExample`

File: catalog/MDCCatalog/MDCNodeListViewController.swift (hunk 1 above)
Comment (yarneo) [reply catalog/MDCCatalog/MDCNodeListViewController.swift]: done

File: catalog/MDCCatalog/MDCNodeListViewController.swift (hunk 2 above)
Comment (yarneo) [reply catalog/MDCCatalog/MDCNodeListViewController.swift]: done

File: components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.h (hunk 3 above)
Comment (yarneo) [reply components/Buttons/examples/supplemental/ButtonsTypicalUseSupplemental.h]: done

File: catalog/MDCCatalog/MDCNodeListViewController.swift (hunk 4 above)
Comment (yarneo) [reply catalog/MDCCatalog/MDCNodeListViewController.swift]: done

File: catalog/MDCCatalog/MDCNodeListViewController.swift [hunk 5]
```diff
@@ -395,54 +418,69 @@ extension MDCNodeListViewController {
...
+      vc = createViewController(from: node)
     } else {
       vc = MDCNodeListViewController(node: node)
     }
     self.navigationController?.pushViewController(vc, animated: true)
   }
+
+  func createViewController(from node: CBCNode) -> UIViewController {
+    var vc: UIViewController
+    let contentVC = selectedNode!.createExampleViewController()
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: Use node here.

File: catalog/MDCCatalog/MDCNodeListViewController.swift [hunk 6]
```diff
@@ -395,54 +418,69 @@ extension MDCNodeListViewController {
...
+      container.appBar.navigationBar.titleAlignment = .center
+      container.appBar.navigationBar.tintColor = UIColor.white
+      container.appBar.navigationBar.titleTextAttributes =
+        [ NSForegroundColorAttributeName: UIColor.white, NSFontAttributeName: appBarFont ]
+      MDCAppBarColorThemer.applySemanticColorScheme(AppTheme.globalTheme.colorScheme,
+                                                    to: container.appBar)
+
+      // TODO(featherless): Remove once
+      // https://github.com/material-components/material-components-ios/issues/367 is resolved.
+      contentVC.title = selectedNode!.title
```
Comment (jverkoey) [inline catalog/MDCCatalog/MDCNodeListViewController.swift]: node

**Review by jverkoey — APPROVED** [review]
LGTM, but please use the arg in the method.

**Pull request merged**